fix(router): preserve replaceUrl when returning a urlTree from CanAct…

blathers16 opened the pull request:
…ivate

This commit will fix the issue of the setting of NavigationExtras.replaceUrl being lost when returning a urlTree from a CanActivateFn.

Fixes #53503

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
     I didn't see anything in the docs that mentions the bug.


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
https://github.com/angular/angular/issues/53503
Issue Number: 53503


## What is the new behavior?
NavigationExtras.replaceUrl is preserved.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

atscott commented on packages/router/src/router.ts:
```diff
@@ -197,8 +197,9 @@ export class Router {
               // if the navigation is rejected.
-              replaceUrl: this.urlUpdateStrategy === 'eager' ||
-                  isBrowserTriggeredNavigation(currentTransition.source)
+              replaceUrl: currentTransition.extras.replaceUrl ??
```
This change isn't quite correct. If `replaceUrl` is specifically `false`, it should still use the other logic, which is meant to capture when the URL has already changed during the current navigation.

blathers16 committed:
fix(router): preserve replaceUrl when returning a urlTree from CanActivate

This commit will fix the issue of the setting of NavigationExtras.replaceUrl being lost when
returning a urlTree from a CanActivateFn.

Fixes #53503

BREAKING CHANGE: When a a guard returns a `UrlTree` as a redirect, the
redirecting navigation will now use `replaceUrl` if the initial
navigation was also using the `replaceUrl` option. If this is not
desirable, the redirect can configure new `NavigationBehaviorOptions` by
returning a `RedirectCommand` with the desired options instead of `UrlTree`.

blathers16 commented:
Hi @atscott I have updated the pull request as well as added extra tests to confirm that your above suggestion is working.

JeanMeche commented:
@blathers16 It looks like something went wrong during rebase. Your PR includes 167 commits, can you have a look ? thank you.

blathers16 commented:
hi @JeanMeche that is probably because I didn't try to rebase.  What is the goal here? To end up with a single commit for the entire PR?

JeanMeche commented:
Yes, the PR should include only the changes intended for the fix. Currently having 1,017 file changes makes it hard to review.

blathers16 commented:
Hi, @JeanMeche, I think I fixed it.  I have no idea where the 1,017 files came from, all I did was merge in upstream/main.  Anyways, it should be all in one commit with no unrelated changes now.

atscott commented on packages/router/src/router.ts:
```diff
@@ -197,8 +197,9 @@ export class Router {
               // if the navigation is rejected.
-              replaceUrl: this.urlUpdateStrategy === 'eager' ||
-                  isBrowserTriggeredNavigation(currentTransition.source)
+              replaceUrl: currentTransition.extras.replaceUrl ? true :
```
```suggestion
              replaceUrl: currentTransition.extras.replaceUrl ||
```

blathers16 replied:
This suggested fix is incorrect logic, it will fail to evaluate the second condition when replaceUrl is false due to short-circuiting.

atscott commented on packages/router/test/computed_state_restoration.spec.ts:
```diff
@@ -11,7 +11,7 @@ import {provideLocationMocks, SpyLocation} from '@angular/common/testing';
 import {ComponentFixture, fakeAsync, TestBed, tick} from '@angular/core/testing';
 import {expect} from '@angular/platform-browser/testing/src/matchers';
-import {Router, RouterModule, RouterOutlet, UrlTree, withRouterConfig} from '@angular/router';
+import {NavigationExtras, Router, RouterModule, RouterOutlet, UrlTree, withRouterConfig} from '@angular/router';
```
Could you please revert these changes and just right a few tests targeting this change directly? The updates in this file make it _really_ difficult to understand and review.

blathers16 replied:
Hi, @atscott, most of those changes were mismatched indentation, (they had stuff indented half a 2-space indent), I fixed that to match the original, can you please take a look and see if it is readable now or if you still need me to simplify it.

atscott replied:
Honestly, I don't think that helps. The tests in this file have a very specific purpose that's _similar_ to the change in this PR but make it unclear what's being tested when reused for a slightly different purpose. I would very much prefer a completely separate set of tests.

blathers16 replied:
@atscott are you saying you would prefer them to be in a separate file?

atscott replied:
Yes, this set of tests was meant to verify `canceledNavigationResolution: 'computed'` by building up several items in the navigation history. It's much more than needed for this change. A new spec file would be easier to review. `regression_integration.spec.ts` could also be a good place. It has a set of tests built around bug fixes and is much shorter than the general dumping ground that is `integration.spec.ts`.

blathers16 replied:
Hi, @atscott I have moved a set of more focused tests into a separate file.  Please review.

atscott reviewed (changes requested):
There's quite a lot of unused code here and tests that could be simplified by using newer features: standalone, functional guards, `RouterTestingHarness`, etc. Older tests in the router code unfortunately aren't good examples of how new ones should be written.

blathers16 commented:
Hi, @atscott, I made most of the requested changes except using RouterTestingHarness.  It doesn't seem to have a way to pass NavigationExtras, so I don't think it can be used in this case.  Feel free to correct me if I am wrong.

atscott commented:
[green TGP](https://fusion2.corp.google.com/presubmit/619277389/OCL:619277389:BASE:619239464:1711527242436:32425d61/targets)

dylhunn commented:
caretaker: this change has a green TGP above

dylhunn commented:
This PR was merged into the repository by commit 60f1d681e0ba66d3d94b0819f2c612f095c2d3d3.
